kubetest2/kubetest2-gke/deployer/deployer.go

```diff
@@ -0,0 +1,681 @@
+
+	// TODO(RonWeber): The version should be plumbed through Extract
+	// or a separate flag rather than magic env variables.
+	if v := os.Getenv("CLUSTER_API_VERSION"); v != "" {
+		args = append(args, "--cluster-version="+v)
+	}
+	args = append(args, d.cluster)
+	fmt.Printf("Gcloud command: gcloud %+v", args)
+	if err := runWithOutput(exec.Command("gcloud", args...)); err != nil {
+		return fmt.Errorf("error creating cluster: %v", err)
+	}
+	for poolName, pool := range shape {
+		if poolName == defaultPool {
+			continue
+		}
+		poolArgs := []string{"node-pools", "create", poolName,
+			"--cluster=" + d.cluster,
+			"--project=" + d.project,
+			loc,
+			"--machine-type=" + pool.MachineType,
+			"--num-nodes=" + strconv.Itoa(pool.Nodes)}
+		poolArgs = append(poolArgs, pool.ExtraArgs...)
+		if err := runWithOutput(exec.Command("gcloud", d.containerArgs(poolArgs...)...)); err != nil {
+			return fmt.Errorf("error creating node pool %q: %v", poolName, err)
+		}
+	}
+	return nil
+}
+
+func (d *deployer) IsUp() (up bool, err error) {
+	// naively assume that if the api server reports nodes, the cluster is up
+	lines, err := exec.CombinedOutputLines(
+		exec.Command("kubectl", "get", "nodes", "-o=name"),
+	)
+	if err != nil {
+		return false, metadata.NewJUnitError(err, strings.Join(lines, "\n"))
+	}
+	return len(lines) > 0, nil
+}
+
+// DumpClusterLogs for GKE generates a small script that wraps
+// log-dump.sh with the appropriate shell-fu to get the cluster
+// dumped.
+//
+// TODO(RonWeber): This whole path is really gross, but this seemed
+// the least gross hack to get this done.
+//
+// TODO(RonWeber): Make this work with multizonal and regional clusters.
+func (d *deployer) DumpClusterLogs() error {
+	// gkeLogDumpTemplate is a template of a shell script where
+	// - %[1]s is the project
+	// - %[2]s is the zone
+	// - %[3]s is a filter composed of the instance groups
+	// - %[4]s is the log-dump.sh command line
+	const gkeLogDumpTemplate = `
+function log_dump_custom_get_instances() {
+  if [[ $1 == "master" ]]; then
+    return 0
+  fi
+
+  gcloud compute instances list '--project=%[1]s' '--filter=%[4]s' '--format=get(name)'
+}
+export -f log_dump_custom_get_instances
+# Set below vars that log-dump.sh expects in order to use scp with gcloud.
+export PROJECT=%[1]s
+export ZONE='%[2]s'
+export KUBERNETES_PROVIDER=gke
+export KUBE_NODE_OS_DISTRIBUTION='%[3]s'
+%[5]s
+`
+	// Prevent an obvious injection.
+	if strings.Contains(d.localLogsDir, "'") || strings.Contains(d.gcsLogsDir, "'") {
+		return fmt.Errorf("%q or %q contain single quotes - nice try", d.localLogsDir, d.gcsLogsDir)
+	}
+
+	// Generate a slice of filters to be OR'd together below
+	if err := d.getInstanceGroups(); err != nil {
+		return err
+	}
+	var filters []string
+	for _, ig := range d.instanceGroups {
+		filters = append(filters, fmt.Sprintf("(metadata.created-by:*%s)", ig.path))
+	}
+
+	// Generate the log-dump.sh command-line
+	var dumpCmd string
+	if d.gcsLogsDir == "" {
```

Comment:
seems this can just be the first case, and then optionally append the second case?

Reply:
Done.